### PR TITLE
#756: Add tests for CustomToolRepository using WireMock

### LINKED ISSUE #756: Implement CustomToolRepositoryTest
https://github.com/devonfw/IDEasy/blob/049cc8c5e709c5493fc310fd76b2414582a7a29e/cli/src/test/java/com/devonfw/tools/ide/repo/CustomToolRepositoryTest.java#L8-L26

This does not look right.
- [ ] We need to create a `custom-tools.json` in `src/test/resources`
- [ ] We need to create a `CustomToolRepository` configured by this `custom-tools.json`
- [ ] Using WireMock and according URL in `custom-tools.json` we could then test downloads of custom tools for that configuration.

### PR DESCRIPTION
**Pull request by vivu001**
### Implemented changes:

* Improve test coverage by adding tests for CustomToolRepository

---

### Checklist for this PR

Make sure everything is checked before merging this PR. For further info please also see
our [DoD](https://github.com/devonfw/IDEasy/blob/main/documentation/contributing/DoD.adoc).

- [x] When running `mvn clean test` locally all tests pass and build is successful
- [x] PR title is of the form `#«issue-id»: «brief summary»` (e.g. `#921: fixed setup.bat`). If no issue ID exists, title only.
- [x] PR top-level comment summarizes what has been done and contains link to addressed issue(s)
- [x] PR and issue(s) have suitable labels
- [x] Issue is set to `In Progress` and assigned to you *or* there is no issue (might happen for very small PRs)
- [x] You followed all [coding conventions](https://github.com/devonfw/IDEasy/blob/main/documentation/coding-conventions.adoc)
- [x] You have added the issue implemented by your PR in [CHANGELOG.adoc](https://github.com/devonfw/IDEasy/blob/main/CHANGELOG.adoc) unless issue is labeled
  with `internal`

### CODE REVIEW

**Comment by Caylipp** [comment]
The test suite looks well structured and comprehensive. Good job 👍
I tested with `mvn clean test` and all tests run successfully!

The Arrange/Act/Assert pattern is used consistently and the scenarios are clearly separated.  
Overall, very clean and solid test coverage.

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -6,6 +6,7 @@ This file documents all notable changes to https://github.com/devonfw/IDEasy[IDE
 
 Release with new features and bugfixes:
 
+* https://github.com/devonfw/IDEasy/issues/756[#756]: Implement CustomToolRepositoryTest
```
Comment (hohwille) [inline CHANGELOG.adoc]: Sorry, the issue was missing the `internal` label, but we only document in CHANGELOG what has impact on the end-user. End-users will not care about our dev internal tests but only about new features and bugfixes.
```suggestion
```

File: cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java [hunk 2]
```diff
@@ -1,28 +1,379 @@
...
+   * @param wmRuntimeInfo wireMock server on a random port.
+   */
+  @Test
+  void testLoadRepositoryFromJson(@TempDir Path tempDir, WireMockRuntimeInfo wmRuntimeInfo) {
+    // arrange
+    setUpRepository(tempDir, wmRuntimeInfo);
+
+    // assert
+    assertThat(this.repository).isNotNull();
+    assertThat(this.repository.getTools()).hasSize(2);
```
Comment (hohwille) [inline cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java]: It is not the best pattern to do
```
var list = ...;
assertThat(list).hasSize(«magic-number»);
```

This is for multiple reasons:
* When reading the test, nobody easily understands such assertion especially when the test-data comes from outside (a different file, a database, etc.) and not within the test-method itself.
* If the test-data changes, the test will break and is not easy to fix properly. If you get `size was 3 but expected 2` or something, then how would you decide if you revealed a bug or the test is broken and the magic number may need to be updated? In many projects this happens and a critical bug may not be fixed because the developer did not understand the context and thought it would be the easiest fix to replace the `2` with a `3` and suddenly the test was green.

However, if you instead do:
```
assertThat(this.repository.getTools().stream().map(tool -> tool.getName() + ":" + tool.getVersion())).contains("jboss-eap:7.1.4.GA", "firefox:70.0.1");
```

This will also allow to search for `jboss-eap` in the project and find it in `src/test/resources/customtools/ide-custom-tools.json`.

File: cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java [hunk 3]
```diff
@@ -1,28 +1,379 @@
...
+  }
+
+  /**
+   * Creates a fully initialized {@link CustomToolRepository} for the current test.
+   *
+   * @param tempDir the temporary test directory used as isolated working directory and parent directory for the generated settings folder.
+   * @param wmRuntimeInfo wireMock server on a random port.
+   */
+  private void setUpRepository(Path tempDir, WireMockRuntimeInfo wmRuntimeInfo) {
+    this.context = new IdeTestContext(tempDir, wmRuntimeInfo);
```
Comment (hohwille) [inline cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java:315]: Are you aware of our test-infrastructure for creating proper instances of `IdeTestContext` in JUnits?
https://github.com/devonfw/IDEasy/blob/main/documentation/contributing/junit-testing.adoc#component-testing

What you are doing here is working but not the proper way to create a reasonable context for e.g. running commandlets.
For this test it is OK and I will resolve this comment.
Before resolving I want you to take notice and if you have not made yourself familiar with our testing infrastructure, please do so...

File: cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java [hunk 4]
```diff
@@ -1,28 +1,379 @@
...
+  }
+
+  /**
+   * Tests that unknown properties in custom-tools.json are silently ignored and do not break loading.
+   *
+   * @param tempDir temporary test directory used as working directory and settings base.
+   * @param wmRuntimeInfo wireMock server on a random port.
+   */
+  @Test
+  void testUnknownPropertiesInJsonAreIgnored(@TempDir Path tempDir, WireMockRuntimeInfo wmRuntimeInfo) {
```
Comment (hohwille) [inline cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java]: how are unknown properties tested here compared to other tests that base on the same `setUpRepository` method?
To be honest this method IMHO tests more or less the same as `testLoadRepositoryFromJson` does.
It is great to have such speaking names in test methods but then you would need to create an extra JSON test file for this test that has unknown properties.
Since there is only one JSON file read by all tests, this does not really make sense and IMHO only causes confusion.

**Review by hohwille — APPROVED** [review]
@vivu001 thanks for your PR. You did a good job to create coverage on this specific edge-case feature of IDEasy. 👍 
I added some comments for further improvements. When resolved, we can merge.

File: cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java (hunk 3 above)
Comment (vivu001) [reply cli/src/test/java/com/devonfw/tools/ide/tool/custom/CustomToolRepositoryTest.java:315]: That was really an oversight. Thanks for pointing it out. I'll definitely look into it.

**Comment by vivu001** [comment]
Successfully retested with version _2026.07.001-07_01_03-SNAPSHOT_.

**Executed:**
```
mvn -pl cli -am clean test
```

**Result:**
SUCCESS on Windows (Git Bash)

Behavior is consistent and works as expected.


```
[INFO] Results:
[INFO]
[WARNING] Tests run: 756, Failures: 0, Errors: 0, Skipped: 18
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] ide dev-SNAPSHOT ................................... SUCCESS [  0.595 s]
[INFO] ide-cli 2026.07.001-SNAPSHOT ....................... SUCCESS [03:37 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:37 min

```
